Civi/Share/CiviMRF/Call.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+namespace Civi\Share\CiviMRF;
+
+use CMRF\Core\AbstractCall;
+use CMRF\Core\Call as CallInterface;
+
+class Call extends AbstractCall implements CallInterface {
+
+  protected string $request_entity;
+
+  protected string $request_action;
+
+  protected array $request;
+
+  protected ?array $reply = NULL;
+
+  protected string $status = CallInterface::STATUS_INIT;
+
+  protected array $metadata = [];
+
+  public static function createNew($connector_id, $core, $entity, $action, $parameters, $options, $callbacks, $factory, string $api_version) {
+    if (!is_array($callbacks)) {
+      if (NULL === $callbacks) {
+        $callbacks = [];
+      }
+      else {
+        $callbacks = [$callbacks];
+      }
+    }
+
+    return static::create($connector_id, $core, $api_version, $entity, $action, $parameters, $options ?? [], $callbacks,
+      $factory
+    );
+  }
+
+  protected static function create(
+    string $connector_id,
+    Core $core,
+    string $api_version,
+    string $entity,
+    string $action,
+    array $parameters,
+    array $options,
+    array $callbacks,
+    CallFactory $factory
+  ): self {
+    $call = new self($core, $connector_id, $factory);
+
+    // compile request
+    if ('3' === $api_version) {
+      $call->request = $call->compileRequest($parameters, $options);
+      $call->request['entity'] = $entity;
+      $call->request['action'] = $action;
+    }
+    elseif ('4' === $api_version) {
+      $call->request = $parameters;
+    }
+    $call->request_entity = $entity;
+    $call->request_action = $action;
+    $call->request['version'] = $api_version;
+    $call->status = CallInterface::STATUS_INIT;
+    $call->metadata['callbacks'] = $callbacks;
+    $call->callbacks = $callbacks;
+
+    return $call;
+  }
+
+  public static function createWithRecord($connector_id, $core, $record, $factory) {
+    // Persisting calls not implemented.
+    return NULL;
+  }
+
+  public function getEntity() {
+    return $this->request_entity;
+  }
+
+  public function getAction() {
+    return $this->request_action;
+  }
+
+  public function getParameters() {
+    return $this->extractParameters($this->request);
+  }
+
+  public function getRequest() {
+    return $this->request;
+  }
+
+  public function getOptions() {
+    return $this->extractOptions($this->request);
+  }
+
+  public function getStatus() {
+    return $this->status;
+  }
+
+  public function setStatus($status, $error_message, $error_code = NULL) {
```

Comment:
This method seems to be relevant only on error so the name isn't appropriate.

Reply:
This will have to be changed upstream in the interface then …